packages/animations/lib/src/utils/composite_animation_widget.dart

```diff
@@ -0,0 +1,253 @@
+/// exit transition.
+///
+/// Used by [CompositeAnimationWidget]
+typedef _ExitTransitionBuilder = Widget Function(
+    BuildContext, Animation<double>, Widget);
+
+/// A widget that animate it's child with different transitions when the
+/// [CompositeAnimationWidget.animation] value goes forward
+/// (built by [CompositeAnimationWidget.enterTransitionBuilder])
+/// and reverse
+/// (built by [CompositeAnimationWidget.exitTransitionBuilder]).
+///
+/// It is used in [FadeThroughTransition], [SharedAxisTransition],
+/// and [FadeScaleTransition] widgets.
+///
+/// This widget also preserve it's child state by building all of it's enter
+/// and exit transitions stacked and only changing animation values of
+/// both transitions instead of dynamically changing widget
+/// subtrees levels in [AnimatedBuilder] based on effective [AnimationStatus]
+/// causes the state object in that particular subtree level to get recreated.
+///
+/// The state preservation of the child will take effect if and only if you're
+/// not changing subtree levels in defined transition builders.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions when
+  /// the [animation] value goes forward
+  /// (built by [enterTransitionBuilder])
+  /// and reverse
+  /// (built by [exitTransitionBuilder]).
+  const CompositeAnimationWidget({
+    @required this.animation,
+    @required this.child,
+    @required this.enterTransitionBuilder,
+    @required this.exitTransitionBuilder,
+    this.flip = false,
+    this.syncVisibleStates = false,
+  })  : assert(child != null),
+        assert(animation != null),
+        assert(enterTransitionBuilder != null),
+        assert(exitTransitionBuilder != null);
+
+  /// The animation that drives the [child]'s entrance and exit.
+  final Animation<double> animation;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// This widget will transition in and out as driven by [animation].
+  final Widget child;
+
+  /// A function that wraps [child] with an animation set define how the
+  /// child appears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes forward.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.completed] (1.0).
+  ///
+  /// Must not change subtree levels.
+  final _EnterTransitionBuilder enterTransitionBuilder;
+
+  /// A function that wraps [child] with an animation set define how the
+  /// child disappears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes reverse.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.dismissed] (0.0).
+  ///
+  /// Must not change subtree levels.
+  final _ExitTransitionBuilder exitTransitionBuilder;
+
+  /// Indicates whether the animation values will be flipped.
```

Comment:
Based on the doc it's not clear to me what this is supposed to do.

Reply:
I just point it out to define builders without conditional building